### ISSUE TITLE
Hi  How to use the Nuget? 

### ISSUE THREAD

**Issue by 599177227** [takuya-takeuchi/OpenPoseDotNet#3]
Hello,I want to ask for instructions for how to use the openpose API,are there some suggestions or instrutions?

**Comment by takuya-takeuchi** [comment]
I'm verry sorry to trouble you.
This product is just development phase. So there are some bug, issue and trouble.
Please forgive me.

1. Add Nuget package to your project
2. Build
3. Copy the following dependency libraries to output directory (Case: Windows)

* boost_chrono-vc140-mt-1_61.dll
* boost_filesystem-vc140-mt-1_61.dll
* boost_python-vc140-mt-1_61.dll
* boost_system-vc140-mt-1_61.dll
* boost_system-vc140-mt-gd-1_61.dll
* boost_thread-vc140-mt-1_61.dll
* caffe.dll
* caffehdf5.dll
* caffehdf5_hl.dll
* caffezlib1.dll
* cublas64_80.dll
* cudart64_80.dll
* cudnn64_5.dll
* curand64_80.dll
* gflags.dll
* gflagsd.dll
* glog.dll
* glogd.dll
* libgcc_s_seh-1.dll
* libgfortran-3.dll
* libopenblas.dll
* libquadmath-0.dll
* opencv_core310.dll
* opencv_ffmpeg310_64.dll
* opencv_imgcodecs310.dll
* opencv_imgproc310.dll
* opencv_world310.dll
* opencv_world310d.dll
* openpose.dll
* python27.dll
* VCRUNTIME140.dll
4. Execute

**Comment by 599177227** [comment]
Oh!!!Thank you very much for your answer
So now We only can execute the Openpose Demo , there is no API available for this NuGet package?

**Comment by takuya-takeuchi** [comment]
> there is no API available for this NuGet package?

What does it mean?
**src/OpenPoseDotNet** wrap openpose.dll. So there should is many openpose API.

OpenPose Demo is ported from https://github.com/CMU-Perceptual-Computing-Lab/openpose/blob/master/examples/openpose/openpose.cpp.

**Comment by 599177227** [comment]
Okay，I just don't know how to use the Openpose API,
Is there a corresponding API document  there or some imformation about it?

**Comment by 599177227** [comment]
And I don‘t Know how to use the OpenposeDonet， Can I use it to implement WPF programs?   I hope that you could give me some instructions?

**Comment by 599177227** [comment]
And May I ask you for some other solutions ,I build the project ,but there are some errors as follwing: 
Error MSB3030: Unable to copy the file "C:\ Users \ 57726 \ source \ repos \ ConsoleApp3 \ packages \ OpenPoseDotNet.1.4.0.20190115 \ runtimes \ win-x64 \ native \ DlibDotNetNativeDnn.dll" because the file could not be found.
My .NetFrame Version:4.72

**Comment by takuya-takeuchi** [comment]
> Is there a corresponding API document there or some imformation about it?
No document. But all api and class have similar with original name. So you could understand easily.

**Comment by takuya-takeuchi** [comment]
> "C:\ Users \ 57726 \ source \ repos \ ConsoleApp3 \ packages \ OpenPoseDotNet.1.4.0.20190115 \ runtimes \ win-x64 \ native \ DlibDotNetNativeDnn.dll" because the file could not be found.

Ops. I'm very sorry to trouble you. I uploaded wrong package.
Please give me time to correct.

**Comment by takuya-takeuchi** [comment]
Sorry. I unlisted package because there are many problems.... Please forgive me......

**Comment by 599177227** [comment]
It doesn't matter, the package you made is already very good. Thank you very much for your answer. Is the error above me because you uploaded the wrong package? Then how should I solve this error?

**Comment by 599177227** [comment]
After I added the two missing files, the program can run, but the console flashes back and then has the following problem in the output box: The exception thrown: "System.Exception" (located in OpenPoseDotNet.dll)
The exception thrown: "System.DllNotFoundException" (located in ConsoleApp1.exe)
Can you help me solve it?

**Comment by takuya-takeuchi** [comment]
Yes, I just fixed now and uploaded 1.4.0.20190117. But nuget package system is difficult and there may be other problems.
Please let me know if you face problem.

Thank you for your cooperation.

**Comment by 599177227** [comment]
Okay，I am grateful for your patient answers。
best wish to you

**Comment by 599177227** [comment]
I am very sorry to trouble you again，I follow the installing steps you give  ，but i meet the problem in the first step，I  failed to open the BuildWindowsVS2015 on Console windows，the problew as folling ：

![problem](https://user-images.githubusercontent.com/45118875/51262024-6f317800-19ec-11e9-90b7-a017d3bb9ca4.png)

**Comment by takuya-takeuchi** [comment]
You must execute in root directory to restore openpose before build OpenPoseDotNet.Native.

````cmd
git submodule update --init
````

**Comment by 599177227** [comment]
Please forgive my ignorance，I don‘t what you meant，how to excute ？and  what is `git submodule updata`  ，I can’t found it on my root directory to restore openpose.

**Comment by takuya-takeuchi** [comment]
ok. Do you install git on windows?

1. move to OpenPose Root dir
````cmd
cd D:\OpenPoseDotNet-1.40.020190117
````
2. execute this command to restore D:\OpenPoseDotNet-1.40.020190117\openpose dir. openpose dir is git submodule directory. This diretory links https://github.com/CMU-Perceptual-Computing-Lab/openpose
````
git submodule update --init
````
3. Build openpose
````
BuildWindowsVS2015.bat
````

**Comment by 599177227** [comment]
And if I build and execute the OpenposeDemo,There are also some problems  as follow:
![image](https://user-images.githubusercontent.com/45118875/51294092-66708e80-1a4c-11e9-8569-60a069d94465.png)
![image](https://user-images.githubusercontent.com/45118875/51294281-370e5180-1a4d-11e9-964e-f5fc5cd23b32.png)

**Comment by takuya-takeuchi** [comment]
WholeBodyFromImage must be specified image file path.

````cmd
dotnet run -c Release -i "c:\test.bmp"
````

And do you install Visual Studio 2015 with C++?
I have never seen this problem.
But if you install visual studio 2017, you can modify BuildVS2015.bat and it is available.

**Comment by 599177227** [comment]
yes ，I  install the visual studio 2017 ,So is there something  i should do to modify BuildVS2015.bat. After I specify the image file path ,there is a  other problem as follwing:
![image](https://user-images.githubusercontent.com/45118875/51325888-e6c4dd00-1aa8-11e9-8ded-0ec311b4c415.png)

**Comment by 599177227** [comment]
yes ，I install the visual studio 2017 ,So is there something I should do to modify BuildVS2015.bat.

**Comment by takuya-takeuchi** [comment]
In bat file,

````bat
cmake -G "Visual Studio 14 2015 Win64" ^
````
to
````bat
cmake -G "Visual Studio 15 2017 Win64" ^
````

**Comment by 599177227** [comment]
OKay ，thank you very much for your answer！

**Comment by 599177227** [comment]
Hello，I am very sorry to trouble you again，I met the problem when I  use the OpenPose NuGet. I bulit the project and reference the NuGet Package,then I copy  All Codes from one example of Your OpenPoseDemo（KeyPointsFromImageConfigurable),in the end I build the prroject , the project will show a console Windows in a flash ,the information was listed below,and I try to use console to build the project, That won't work.
the detail information as following：
![image](https://user-images.githubusercontent.com/45118875/51422274-cf8c0980-1be6-11e9-9546-5e4ec60208cd.png)
![image](https://user-images.githubusercontent.com/45118875/51422276-d9ae0800-1be6-11e9-908a-575aaa355730.png)

I don‘t know how to make it,Could you help me out ?
Thank you very much!!

**Comment by takuya-takeuchi** [comment]
You 'ONLY' copy demo project to other directory?
If so, you must modify ****.csporj.

Because sample project refers OpenPoseDotNet project by relative path.
So copied project could not read OpenPoseDotNet project.

And could you translate error message to english or paste error message here?
I cannot understand Chinese and input it to translater because I don't know how to read them.
(I am Japanese. But some Chinese words are not used in Japan. so I cannot input!!)

**Comment by 599177227** [comment]
Sure，The red font are translated to follow： 
Wúfǎ yùnxíng xiàngmù
qǐng quèbǎo nǐ jùyǒu kě yùnxíng de xiàngmù lèixíng qiě “dotnet run” zhīchí cǐ xiàngmù.
Kě yùnxíng de xiàngmù yīng miànxiàng kě yùnxíng de TFM(lìrú netcoreapp2.0) Qiě qí OutputType wèi “Exe“.
Dāngqián de OutputType wèi”Exe“
展开
114/5000
Unable to run project
Make sure you have a working project type and "dotnet run" supports this project.
A runnable project should target a runnable TFM (for example, netcoreapp2.0) and its OutputType is "Exe".
The current OutputType is "Exe"

I don't  exactly what you meant,my root directory don’t have any ****.csporj file.

Could you Download My project  in my reposiyories，and try to run it.

**Comment by 599177227** [comment]
I don't exactly know what you meant,my root directory don’t have any ****.csporj file.

**Comment by 599177227** [comment]
If my expression makes you unable to understand, please forgive my limited English level.

**Comment by takuya-takeuchi** [comment]
> If my expression makes you unable to understand, please forgive my limited English level.

Don't worry. My english is also not good. :)

> Could you Download My project in my reposiyories，and try to run it.

OK. Where is your repository? Is it correct ?
https://github.com/599177227/FacialDetected

**Comment by 599177227** [comment]
No ，that is wrong ，the file which is named OpenposeBody   is uploading ，please wait .

**Comment by 599177227** [comment]
Now it is Ok, you can download at the Release Page in Openpose .please help me  check it ，thank you ！

**Comment by takuya-takeuchi** [comment]
OK.
OpenposeBody is created as .NET Framework 4.7.2 project.
But our sample is created as .NET Core console app.

So dotnet commands does not work for OpenposeBody.
You should re-create project as .NET Core console app if you want to use dotnet command.

**Comment by 599177227** [comment]
Thank you ，I have received it，I will give it a try.

**Comment by 599177227** [comment]
Hello,Could I ask you another problem?I want to know the OpenPose API  <openpose/flags.hpp>is what?
why I can't found it? Could you tell me ?
![image](https://user-images.githubusercontent.com/45118875/51427137-bd35be00-1c2e-11e9-9433-5f785aa9ccdb.png)
I am sorry to bother you again！

**Comment by 599177227** [comment]
Hello, can I ask a question about the C++ API used in C#? I used the openpose source code and compiled the first API dll. I introduced it to my C# code, but it reported this error. Have you ever met it? I don't know what is configured less?
can you help me?
![image](https://user-images.githubusercontent.com/45118875/51437917-da769500-1cdf-11e9-80f9-ff5c80ae60b9.png)
![image](https://user-images.githubusercontent.com/45118875/51437923-e8c4b100-1cdf-11e9-9db2-bc7e8957db39.png)
![image](https://user-images.githubusercontent.com/45118875/51437925-f417dc80-1cdf-11e9-935f-3193b5543a9b.png)

**Comment by 599177227** [comment]
thank you sincerely！

**Comment by takuya-takeuchi** [comment]
OpenPose needs Caffe model files.
This files are generated when building openpose. (Not building OpenPoseDotNet!!!!)
Please check OpenPoseDotNet\openpose. You could see model directory.

**Comment by 599177227** [comment]
Thank you very much for solving a problem that I have not been able to solve. Thank you very much!

**Comment by 599177227** [comment]
Hi，friend，there is another problem  bothered me thoese days，Why my project sometimes run so well，but sometime it will occur error and I can‘t fit it，I don’t know why and how to solve it ，could you give me some instrutions about it ，the problem  as following ：
![image](https://user-images.githubusercontent.com/45118875/51472770-64dcf880-1db5-11e9-88bd-a4ca2dacc6a3.png)

**Comment by takuya-takeuchi** [comment]
Is it OpenPoseDotNet sample?
It looks like to be https://github.com/CMU-Perceptual-Computing-Lab/openpose/blob/master/examples/tutorial_api_cpp/01_body_from_image_default.cpp.
I have no idea to resolve issue.

But it indicates https://github.com/CMU-Perceptual-Computing-Lab/openpose/blob/master/include/openpose/core/wScaleAndSizeExtractor.hpp#L68.
It seems input image becomes sometimes empty size. It may be OpenCV issue.

**Comment by 599177227** [comment]
No,it is not a sample of OpenPoseDoNet,I only want to wrapper the C++ to the dll for C#,but if I set the fuction return a void ,this is ok, if I set it to return cv::Mat,that is a Unknow problem,the detail  codes as following ：
![image](https://user-images.githubusercontent.com/45118875/51514160-8091de80-1e49-11e9-88dc-50096107df09.png)
![Uploading image.png…]()

I now implement a painting of the image by calling a C++ .dll file and saving it to a local path, but I can't return cv::mat type or byte* directly to C#, so I can only wait for it to finish painting. Local reading will increase the running time. Do you have any way to solve it?

**Comment by takuya-takeuchi** [comment]
> System.DllNotFoundException

It means there is not 01_body_From_image_default.dll or it's depenecies lib.
How about using dependency walker to inspect missing library?

**Comment by YuBin8** [comment]
> boost_chrono-vc140-mt-1_61.dll
boost_filesystem-vc140-mt-1_61.dll
boost_python-vc140-mt-1_61.dll
boost_system-vc140-mt-1_61.dll
boost_system-vc140-mt-gd-1_61.dll
boost_thread-vc140-mt-1_61.dll
caffe.dll
caffehdf5.dll
caffehdf5_hl.dll
caffezlib1.dll
cublas64_80.dll
cudart64_80.dll
cudnn64_5.dll
curand64_80.dll
gflags.dll
gflagsd.dll
glog.dll
glogd.dll
libgcc_s_seh-1.dll
libgfortran-3.dll
libopenblas.dll
libquadmath-0.dll
opencv_core310.dll
opencv_ffmpeg310_64.dll
opencv_imgcodecs310.dll
opencv_imgproc310.dll
opencv_world310.dll
opencv_world310d.dll
openpose.dll
python27.dll
VCRUNTIME140.dll

where are these file?

**Comment by takuya-takeuchi** [comment]
@YuBin8 
You have to build.
Please check https://github.com/takuya-takeuchi/OpenPoseDotNet#1-build

**Comment by YuBin8** [comment]
@takuya-takeuchi  thanks, but when i run build , it say  
Cant Not Found Type [Config]。
+ $Config = [Config]::new($OpenPoseDotNetRoot, $Configuration, $Target, ...
+           ~~~~~~~~
    + CategoryInfo          : InvalidOperation: (Config:TypeName) []，RuntimeException
    + FullyQualifiedErrorId : TypeNotFound

**Comment by takuya-takeuchi** [comment]
@YuBin8 
Did you used powershell core (or 7) rather than built-in powershell?
I can not reproduce it.

**Comment by YuBin8** [comment]
> @YuBin8 Did you used powershell core (or 7) rather than built-in powershell? I can not reproduce it.

thank you so much for your patience,i was used built-in powershell, now when i run it in powershell core  7 sitll has some err



![image](https://user-images.githubusercontent.com/7712016/204477923-96a30740-8a57-4663-83b5-c0cfca14b52e.png)

my vs version :  Microsoft Visual Studio Community 2019           16.11.19
vc++ version 
![image](https://user-images.githubusercontent.com/7712016/204483440-a2601cd7-a154-47ff-b7db-a493805e416d.png)

**Comment by takuya-takeuchi** [comment]
@YuBin8 
You must install Cmake from https://cmake.org/download/

**Comment by YuBin8** [comment]
@takuya-takeuchi 
so sorry to bothering you , but i still can't get it  build
now it say 


  Checking Build System
  Building Custom Rule D:/OpenPoseDotNet-master/src/OpenPoseDotNet.Native/CMakeLists.txt
  CameraParameterReader.cpp
D:\OpenPoseDotNet-master\src\OpenPoseDotNet.Native\openpose\3d\../shared.h(4,10): fatal error C1083: Cannot open include file: “opencv2/opencv.hpp”: No such fi
le or directory [D:\OpenPoseDotNet-master\src\OpenPoseDotNet.Native\build_win_desktop_cpu_x64_d\OpenPoseDotNetNative.vcxproj]
  MatExpr.cpp
D:\OpenPoseDotNet-master\src\OpenPoseDotNet.Native\openpose\3rdparty\opencv\../../shared.h(4,10): fatal error C1083: Cannot open include file: “opencv2/opencv.
hpp”: No such file or directory [D:\OpenPoseDotNet-master\src\OpenPoseDotNet.Native\build_win_desktop_cpu_x64_d\OpenPoseDotNetNative.vcxproj]
  cv.cpp


i added system environment variables path for opencv2   still don't work

**Comment by takuya-takeuchi** [comment]
@YuBin8 
I tried to build from first step on my dev machine.

````bat
$ git clone https://github.com/takuya-takeuchi/OpenPoseDotNet
$ cd OpenPoseDotNet
$ git checkout develop
$ git submodule update --init --recursive
$ cd src\OpenPoseDotNet.Native
# you need not to add last argment 101 if you does not use cuda
# And Debug build requires huge memory for compile so you could be better to use Release
$ pwsh Build.ps1 Debug cpu 64 desktop 2019
Build OpenPoseDotNet.Native

    Directory: E:\OpenPoseDotNet\src\OpenPoseDotNet.Native

Mode                 LastWriteTime         Length Name
----                 -------------         ------ ----
d----          2022/11/30    20:29                build_win_desktop_cpu_x64_d
Start Build openpose
   cmake -G Visual Studio 16 2019 -A x64 -D CMAKE_BUILD_TYPE=Debug
         -D USE_CUDA:BOOL=OFF
         -D GPU_MODE=CPU_ONLY
         E:\OpenPoseDotNet\src\openpose
   cmake --build . --config Debug
End Build openpose
Start Build OpenPoseDotNet.Native
   cmake -G  Visual Studio 16 2019 -A x64 -T host=x64
         -D BUILD_SHARED_LIBS=ON
         ..
-- Selecting Windows SDK version 10.0.22000.0 to target Windows 10.0.19044.
-- The C compiler identification is MSVC 19.29.30147.0
-- The CXX compiler identification is MSVC 19.29.30147.0
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Check for working C compiler: C:/Program Files (x86)/Microsoft Visual Studio/2019/Community/VC/Tools/MSVC/14.29.30133/bin/Hostx64/x64/cl.exe - skipped
-- Detecting C compile features
-- Detecting C compile features - done
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Check for working CXX compiler: C:/Program Files (x86)/Microsoft Visual Studio/2019/Community/VC/Tools/MSVC/14.29.30133/bin/Hostx64/x64/cl.exe - skipped
-- Detecting CXX compile features
-- Detecting CXX compile features - done
-------------------------------------------------------
-- CMAKE_SYSTEM_INFO_FILE: Platform/Windows
-- CMAKE_SYSTEM_NAME:      Windows
-- CMAKE_SYSTEM_PROCESSOR: AMD64
-- CMAKE_SYSTEM:           Windows-10.0.19044
-- CMAKE_C_COMPILER:       C:/Program Files (x86)/Microsoft Visual Studio/2019/Community/VC/Tools/MSVC/14.29.30133/bin/Hostx64/x64/cl.exe
-- CMAKE_CXX_COMPILER:     C:/Program Files (x86)/Microsoft Visual Studio/2019/Community/VC/Tools/MSVC/14.29.30133/bin/Hostx64/x64/cl.exe
-- os_version_suffix:
-- Project: E:/OpenPoseDotNet/src/OpenPoseDotNet.Native
-- Binaries: E:/OpenPoseDotNet/src/OpenPoseDotNet.Native/build_win_desktop_cpu_x64_d
-------------------------------------------------------
CMake Warning (dev) at CMakeLists.txt:30 (if):
  Policy CMP0054 is not set: Only interpret if() arguments as variables or
  keywords when unquoted.  Run "cmake --help-policy CMP0054" for policy
  details.  Use the cmake_policy command to set the policy and suppress this
  warning.

  Quoted variables like "MSVC" will no longer be dereferenced when the policy
  is set to NEW.  Since the policy is not set the OLD behavior will be used.
This warning is for project developers.  Use -Wno-dev to suppress it.

-- Configuring done
-- Generating done
-- Build files have been written to: E:/OpenPoseDotNet/src/OpenPoseDotNet.Native/build_win_desktop_cpu_x64_d
.NET Framework 向け Microsoft (R) Build Engine バージョン 16.11.2+f32259642
Copyright (C) Microsoft Corporation.All rights reserved.

  Checking Build System
  Building Custom Rule E:/OpenPoseDotNet/src/OpenPoseDotNet.Native/CMakeLists.txt
  CameraParameterReader.cpp
  MatExpr.cpp
  VideCapture.cpp
  cv.cpp
  mat.cpp
  Matrix.cpp
  cvMatToOpInput.cpp
  cvMatToOpOutput.cpp
  datum.cpp
  macros.cpp
  opOutputToCvMat.cpp
  point.cpp
  rectangle.cpp
  scaleAndSizeExtractor.cpp
  custom.cpp
  customDatum.cpp
  fileStream.cpp
  flags.cpp
  gpu.cpp
  frameDisplayer.cpp
  コードを生成中...
  コンパイル中...
  gui.cpp
  wGui.cpp
  poseCpuRenderer.cpp
  poseExtractorCaffe.cpp
  poseExtractorNet.cpp
  poseGpuRenderer.cpp
  poseParameters.cpp
  poseRenderer.cpp
  datumProducer.cpp
  producer.cpp
  videoCaptureReader.cpp
  wDatumProducer.cpp
  webcamReader.cpp
  time_point.cpp
  shared_ptr.cpp
  vector.cpp
  threadManager.cpp
  user.cpp
  userWorker.cpp
  userWorkerConsumer.cpp
  コードを生成中...
  コンパイル中...
  userWorkerProducer.cpp
  check.cpp
  errorAndLog.cpp
  flagsToOpenPose.cpp
  profiler.cpp
  wrapper.cpp
  wrapperAuxiliary.cpp
  wrapperStructExtra.cpp
  wrapperStructFace.cpp
  wrapperStructGui.cpp
  wrapperStructHand.cpp
  wrapperStructInput.cpp
  wrapperStructOutput.cpp
  wrapperStructPose.cpp
  コードを生成中...
  array.cpp
  string.cpp
  array.cpp
  string.cpp
     ライブラリ E:/OpenPoseDotNet/src/OpenPoseDotNet.Native/build_win_desktop_cpu_x64_d/Debug/OpenPoseDotNetNative.lib とオブジェクト E:/OpenPoseDotNet/src/OpenPoseDotNet.Native/build_win_desktop_cpu_x64_d/Debug/OpenP
  oseDotNetNative.exp を作成中
  OpenPoseDotNetNative.vcxproj -> E:\OpenPoseDotNet\src\OpenPoseDotNet.Native\build_win_desktop_cpu_x64_d\Debug\OpenPoseDotNetNative.dll
  Building Custom Rule E:/OpenPoseDotNet/src/OpenPoseDotNet.Native/CMakeLists.txt
Copy PreBuild openpose binaries
Copy openpose binary
````

There is no issue.
